File: neetcode/blind75/graphs/course_schedule.py
# https://neetcode.io/problems/course-schedule

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # numCourses != len(prerequisites)
        preMap = {i: [] for i in range(numCourses)}

        for crs, pre in prerequisites:
            preMap[crs].append(pre)

        visited = set()

        def dfs(course):
            if course in visited:
                return False
            if preMap[course] == []:
                # no preReq
                return True

            visited.add(course)

            # eval for loop within this course's prereqs
            for pre in preMap[course]:
                if not dfs(pre):
                    return False

            # clear visited so next course can use
            visited.remove(course)
            # clear prereq for this course so if another course prereq uses this course, it does not repeat checking this course's prereq.
            preMap[course] = []
            return True
            
        for course in range(numCourses):
            if not dfs(course):
                return False

        return True

File: neetcode/neet150_JS/Graphs/course-schedule.js
// https://neetcode.io/problems/course-schedule

class Solution {
    /**
     * @param {number} numCourses
     * @param {number[][]} prerequisites
     * @return {boolean}
     */
    canFinish(numCourses, prerequisites) {
        if (numCourses === 0 || prerequisites.length === 0) {
            return true
        }

        const visited = new Set()

        const adjList = Array(numCourses).fill().map((e) => {return Array(0)})
        for (let [crs, pre] of prerequisites) {
            adjList[crs].push(pre)
        }

        for (let i = 0; i < numCourses; i ++) {
            if (!this.DFS(i, adjList, visited)) {
                return false
            }
        }

        return true
    }

    DFS(crs, adjList, visited) {
        if (visited.has(crs)) {
            return false
        }

        visited.add(crs)

        for (let pre of adjList[crs]) {
            if (!this.DFS(pre, adjList, visited)) {
                return false
            }
        }

        visited.delete(crs)

        return true
    }
}
